### DISCUSSION TITLE
Clarification on which Bootstrap imports are required vs. optional for Sass compilation in help docs

### DISCUSSION THREAD

**Discussion by picobots** [twbs/bootstrap#39228] [Q&A] [answered]
In the [help documentation](https://getbootstrap.com/docs/5.3/customize/sass/) for compiling Bootstrap with Sass, the ordering of Bootstrap imports is slightly different than the ordering of the imports in the actual [Bootstrap framework source](https://github.com/twbs/bootstrap/blob/v5.3.2/scss/bootstrap.scss).

This is how the imports are structured in the actual framework:

```
// scss-docs-start import-stack
// Configuration
@import "functions";
@import "variables";
@import "variables-dark";
@import "maps";
@import "mixins";
@import "utilities";      // <-- note utilities order

// Layout & components
@import "root";           // <-- note root order
@import "reboot";
@import "type";
// etc.
```

However, in the Sass documentation, the order of imports is different—specifically `root` and `utilities` are reversed throughout the examples on the page:

```
// 1. Include functions first (so you can manipulate colors, SVGs, calc, etc)
@import "../node_modules/bootstrap/scss/functions";

// 2. Include any default variable overrides here

// 3. Include remainder of required Bootstrap stylesheets (including any separate color mode stylesheets)
@import "../node_modules/bootstrap/scss/variables";
@import "../node_modules/bootstrap/scss/variables-dark";

// 4. Include any default map overrides here

// 5. Include remainder of required parts
@import "../node_modules/bootstrap/scss/maps";
@import "../node_modules/bootstrap/scss/mixins";
@import "../node_modules/bootstrap/scss/root";            // <-- note root order

// 6. Optionally include any other parts as needed
@import "../node_modules/bootstrap/scss/utilities";       // <-- note utilities order
@import "../node_modules/bootstrap/scss/reboot";
@import "../node_modules/bootstrap/scss/type";
```

Generally speaking, for compiling Bootstrap from Sass, I'm guessing this is fine and it doesn't matter if `root` and `utilities` are swapped.

However, to keep my Sass compilation time lightning fast in my projects, I like to compile Bootstrap independently of my main site Sass (so that Bootstrap isn't recompiled every time I make a small change to my main site Sass). But I still need access to Bootstrap functions/variables/etc. in my main site Sass, so the question is: **Where should I _stop_ including Bootstrap imports if I just want access to Bootstrap's functions/variables/etc. in my main site Sass, but I don't want any actual Bootstrap CSS rules in my main site CSS?**

Is this the right set of imports?

```
@import "functions";
@import "variables";
@import "variables-dark";
@import "maps";
@import "mixins";
@import "utilities";
```

Or is this the right set of imports?

```
@import "functions";
@import "variables";
@import "variables-dark";
@import "maps";
@import "mixins";
@import "root";
```

Is `root` considered optional, or is `utilities` considered optional, or both or neither?

**Comment by julien-deramond** [comment]
* `root` is required: it's where you can find low-level custom properties defined in `:root` (e.g. `--bs-body-color`, `--bs-body-bg`, etc.)
* `utilities` is only needed if you also import `utilities/api` which generates utilities (e.g. `.m-0`, `.p-3`, `.position-fixed`, etc.). So if you don't use Bootstrap utilities, it's not needed.
* `reboot` is labeled as optional but can be handy depending on your needs (see https://getbootstrap.com/docs/5.3/content/reboot/)
* the rest depends on what you actually use in your project so optional

**Comment by picobots** [comment, marked as answer]
For what it's worth, I did some testing, and for folks like me who are compiling Bootstrap **separately** (e.g. as a second standalone entrypoint in Webpack), if I import `root` or `reboot` again in my main site Sass (or anything past `utilities` in [the framework bootstrap.scss file](https://github.com/twbs/bootstrap/blob/v5.3.2/scss/bootstrap.scss) for that matter), then that results in duplicated Bootstrap CSS in my compiled main site CSS, and it also results in significantly slower Sass compilation time, even if I only make a change in my main site Sass.

To clarify, this is my scenario:

```
// Entry for Webpack bundling process
entry: [
    // Full Bootstrap framework Sass, with my variable customization
    "./assets/src/styles/bootstrap/bootstrap.scss",
    // Main site Sass - this should have access Bootstrap Sass variables, but shouldn't have ANY compiled Bootstrap framework CSS
    "./assets/src/styles/main.scss"
]
```

If I import `root` in `main.scss`, then I get duplicated Bootstrap `:root` variables in my compiled main site CSS (because they were already compiled and included by the first `bootstrap` Webpack entrypoint). If I stop at `utilities`, that seems to be the sweet spot—then I have full access to all of the Bootstrap Sass vars/maps in my main Sass, but none of the core Bootstrap CSS gets duplicated.

So it seems like this is the correct set up for Bootstrap imports when you just want access to the Bootstrap Sass variables/maps but don't want any actual Bootstrap CSS:

```
// Bootstrap functions
@import "~bootstrap/scss/functions";

// Bootstrap variable customization
@import "bootstrap-variables";

// Bootstrap variable defaults
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/variables-dark";

// Bootstrap map customization
@import "bootstrap-maps";

// Remaining Bootstrap imports
@import "~bootstrap/scss/maps";
@import "~bootstrap/scss/mixins";
@import "~bootstrap/scss/utilities";

// DO NOT IMPORT ANY OF THESE
// @import "~bootstrap/scss/root";
// @import "~bootstrap/scss/reboot";
// @import "~bootstrap/scss/type";
```
